x-pack/platform/plugins/shared/alerting/server/alerts_client/alerts_client.ts

```diff
@@ -179,67 +178,56 @@ export class AlertsClient<
 
-    const { runTimestamp, trackedExecutions } = opts;
+    const { runTimestamp } = opts;
 
     if (runTimestamp) {
       this.runTimestampString = runTimestamp.toISOString();
     }
     await this.legacyAlertsClient.initializeExecution(opts);
 
-    this.trackedExecutions = new Set(trackedExecutions ?? []);
-
     // No need to fetch the tracked alerts for the non-lifecycle rules
     if (this.ruleType.autoRecoverAlerts) {
-      const getTrackedAlertsByExecutionUuids = async (executionUuids: string[]) => {
-        const result = await this.search({
-          size: (opts.maxAlerts || DEFAULT_MAX_ALERTS) * 2,
-          seq_no_primary_term: true,
+      const getTrackedAlerts = async () => {
+        // We can use inner_hits to get the alerts for the most recent executions
+        // but this may return too many alerts, therefore we make two queries:
+        // 1. Get the most recent execution UUIDs
+        // 2. Get the alerts for those execution UUIDs
+        const executions = await this.search({
+          size: opts.flappingSettings.lookBackWindow,
           query: {
             bool: {
               must: [{ term: { [ALERT_RULE_UUID]: this.options.rule.id } }],
-              filter: [{ terms: { [ALERT_RULE_EXECUTION_UUID]: executionUuids } }],
+              must_not: [{ term: { [ALERT_STATUS]: ALERT_STATUS_UNTRACKED } }],
             },
           },
+          collapse: {
+            field: ALERT_RULE_EXECUTION_UUID,
+          },
+          _source: false,
+          sort: [{ [TIMESTAMP]: { order: 'desc' } }],
         });
-        return result.hits;
-      };
-
-      const getTrackedAlertsByAlertUuids = async () => {
-        const { activeAlertsFromState = {}, recoveredAlertsFromState = {} } = opts;
-        const uuidsToFetch: string[] = [];
-        Object.values(activeAlertsFromState).forEach((activeAlert) =>
-          uuidsToFetch.push(activeAlert.meta?.uuid!)
-        );
-        Object.values(recoveredAlertsFromState).forEach((recoveredAlert) =>
-          uuidsToFetch.push(recoveredAlert.meta?.uuid!)
-        );
 
-        if (uuidsToFetch.length <= 0) {
-          return [];
-        }
+        const executionUuids = executions.hits
+          .map((hit) => get(hit.fields, ALERT_RULE_EXECUTION_UUID))
+          .flat();
 
-        const result = await this.search({
-          size: uuidsToFetch.length,
+        const alerts = await this.search({
+          size: (opts.maxAlerts || DEFAULT_MAX_ALERTS) * 2,
           seq_no_primary_term: true,
-          sort: { [ALERT_START]: 'desc' },
           query: {
             bool: {
-              filter: [
-                { term: { [ALERT_RULE_UUID]: this.options.rule.id } },
-                { terms: { [ALERT_UUID]: uuidsToFetch } },
-              ],
+              must: [{ term: { [ALERT_RULE_UUID]: this.options.rule.id } }],
+              filter: [{ terms: { [ALERT_RULE_EXECUTION_UUID]: executionUuids } }],
```

Comment:
do you think we need to exclude status: untracked in this query? we didn't before, but I think that might have been an oversight?

Reply:
Oh! thanks for pointing out. I overlooked it.
Actually the filter should be here, having it in the other one may cause to skip an execution.
Done.